curp/src/client_new/retry.rs

```diff
@@ -0,0 +1,256 @@
+#![allow(unused)] // TODO: remove
+
+use std::{marker::PhantomData, ops::SubAssign, sync::Arc, time::Duration};
+
+use async_trait::async_trait;
+use curp_external_api::cmd::Command;
+use futures::Future;
+use tracing::warn;
+
+use crate::{
+    members::ServerId,
+    rpc::{
+        connect::ConnectApi, ConfChange, CurpError, FetchClusterResponse, Member, ReadState,
+        Redirect,
+    },
+};
+
+use super::{ClientApi, LeaderStateUpdate, ProposeResponse};
+
+/// Backoff config
+#[derive(Debug, Clone)]
+pub(super) enum BackoffConfig {
+    /// A fixed delay backoff
+    Fixed,
+    /// A exponential delay backoff
+    Exponential {
+        /// Control the max delay of exponential
+        max_delay: Duration,
+    },
+}
+
+/// Retry config to control the retry policy
+#[derive(Debug, Clone)]
+pub(super) struct RetryConfig {
+    /// Backoff config
+    backoff: BackoffConfig,
+    /// Initial delay
+    delay: Duration,
+    /// Retry count
+    count: usize,
+    /// Enable jitter to randomize the delay to avoid thundering herd
```

Comment:
Why should we handle thundering herd on the client side? IMO, the thundering herd occurs when a large number of processes or threads waiting for an event are awoken when that event occurs, but only one process will finally be able to handle that event. The tonic server will spawn the relavent task to handle the rpc call from client. These rpc calls needn't wait for the same event.

Reply:
> Why should we handle thundering herd on the client side?

The CURP client can also be on the server.

I think it is meaningful, at the same time, a large number of retry requests occurring can also impose significant pressure on the entire system. (Perhaps the opportunity for a large number of retries is because the quantity of requests is too large for the system to handle. In this case, jitter can be used to introduce some time variance to these requests during the next retry, allowing the system to break out of this loop.)

Reply:
> only one process

Here, the `only one process` refers to the entire system.

Reply:
But it's not the thundering herd problem. Please modify the comment.

Reply:
It is thundering herd problem

Reply:
@rogercloud

Reply:
I have a little question. I think nodes in a cluster should normally be deployed on different servers (I'm not sure if this is correct). Each node will receive requests from different clients. These different clients don't seem to do this because of differences in send times and network latency etc.

Reply:
removed